x-pack/legacy/plugins/siem/server/lib/detection_engine/alerts/utils.test.ts

```diff
@@ -0,0 +1,658 @@
+      );
+      expect(mockLogger.error).toHaveBeenCalledTimes(2);
+      expect(result).toEqual(false);
+    });
+    test('if unsuccessful iteration of searchAfterAndBulkIndex due to empty sort ids', async () => {
+      const searchAfterSortId = '1234567891111';
+      const sampleParams = sampleSignalAlertParams(undefined);
+      const savedObjectsClient = savedObjectsClientMock.create();
+      const expectedSearchAfterQuery = buildEventsSearchQuery({
+        index: sampleParams.index,
+        from: sampleParams.from,
+        to: sampleParams.to,
+        filter: sampleParams.filter,
+        size: sampleParams.size ? sampleParams.size : 1,
+        searchAfterSortId,
+        maxDocs: undefined,
+      });
+      const mockService: AlertServices = {
+        callCluster: async (action: string, params: SearchParams) => {
+          if (action === 'bulk') {
+            expect(action).toEqual('bulk');
+            expect(params.index).toEqual('.siem-signals-10-01-2019');
+            return {
+              took: 100,
+              errors: false,
+              items: [
+                {
+                  fakeItemValue: 'fakeItemKey',
+                },
+              ],
+            };
+          }
+          expect(action).toEqual('search');
+          expect(params.index).toEqual(sampleParams.index);
+          expect(params).toEqual(expectedSearchAfterQuery);
+          return sampleDocSearchResultsWithSortId;
+        },
+        alertInstanceFactory: jest.fn(),
+        savedObjectsClient,
+      };
+      const result = await searchAfterAndBulkIndex(
+        sampleDocSearchResultsNoSortId,
+        sampleParams,
+        mockService,
+        mockLogger
+      );
+      expect(mockLogger.warn).toHaveBeenCalledTimes(1);
+      expect(result).toEqual(false);
+    });
+    test('if unsuccessful iteration of searchAfterAndBulkIndex due to empty sort ids and 0 total hits', async () => {
+      const searchAfterSortId = '1234567891111';
+      const sampleParams = sampleSignalAlertParams(undefined);
+      const savedObjectsClient = savedObjectsClientMock.create();
+      const expectedSearchAfterQuery = buildEventsSearchQuery({
+        index: sampleParams.index,
+        from: sampleParams.from,
+        to: sampleParams.to,
+        filter: sampleParams.filter,
+        size: sampleParams.size ? sampleParams.size : 1,
+        searchAfterSortId,
+        maxDocs: undefined,
+      });
+      const mockService: AlertServices = {
+        callCluster: async (action: string, params: SearchParams) => {
+          if (action === 'bulk') {
+            expect(action).toEqual('bulk');
+            expect(params.index).toEqual('.siem-signals-10-01-2019');
+            return {
+              took: 100,
+              errors: false,
+              items: [
+                {
+                  fakeItemValue: 'fakeItemKey',
+                },
+              ],
+            };
+          }
+          expect(action).toEqual('search');
```

Comment:
Per my earlier comment about trying to have only 1, maybe 2 `expect` statement at most per test for maintainers and developers to easily understand what is going on and reading code...

This is a bit of an example of why that is needed because starting on lines 494 thorough 497, these expect statements are never reached. Reason being is that on line 484, you have an early return statement.

This part of the test is confusing to me because I see one expect statement which says:

```ts
expect(action).toEqual('bulk');
```

But then on 494 I see a contradiction of:

```ts
expect(action).toEqual('search');
```

Since you have early return statements for this and if we followed the paradigm of the least amount of expect statements per test this would actually become this code:

```ts
    test('if one successful iteration of searchAfterAndBulkIndex', async () => {
      const searchAfterSortId = '1234567891111';
      const sampleParams = sampleSignalAlertParams(undefined);
      const savedObjectsClient = savedObjectsClientMock.create();
      buildEventsSearchQuery({
        index: sampleParams.index,
        from: sampleParams.from,
        to: sampleParams.to,
        filter: sampleParams.filter,
        size: sampleParams.size ? sampleParams.size : 1,
        searchAfterSortId,
        maxDocs: undefined,
      });
      const mockService: AlertServices = {
        callCluster: async (action: string, params: SearchParams) => {
          expect(action).toEqual('bulk');
          expect(params.index).toEqual('.siem-signals-10-01-2019');
          return {
            took: 100,
            errors: false,
            items: [
              {
                fakeItemValue: 'fakeItemKey',
              },
            ],
          };
        },
        alertInstanceFactory: jest.fn(),
        savedObjectsClient,
      };
      const result = await searchAfterAndBulkIndex(
        sampleDocSearchResultsWithSortId,
        sampleParams,
        mockService,
        mockLogger
      );
      expect(result).toEqual(true);
    });
```

You do want to avoid dead branching logic within tests as much as possible such as if statements or larger harnesses so we can be sure that each `expect` statement is reached and each `expect` statement does indeed do what we intend it to do which is prove a particular value's truthfulness.